### PR TITLE
Fix/wallet for new accounts

### LINKED ISSUE #8760: Added account is duplicated when multutapping Finish button
# Problem

If i multitap [Finish] button when adding an account to my multiaccount instance -> then added account will be duplicated in wallet

image

## Steps

- create new account
- tap wallet -> add account
- tap Add an account
- tap Generate a new account
- enter password -> generate new account
- select account color
- tap [Finish] button several times in a row
- check wallet

## Video

https://drive.google.com/open?id=1Xa0j7wG0StB22sSOvRHFgzO9yaHTxA0I

## Notes

Similar to https://github.com/status-im/status-react/issues/8710 

@flexsurfer @siphiuel

### LINKED ISSUE #8802: No redirect from universal links after login
#  Description

*Type*: Bug

*Summary*: no redirection to user profile / public chat / Dapp if you open the link and should login.
On 0.13.0 if you open link `https://get.status.im/user/0x0449add8977ca4aea3c5b7dde3cfca4a1e59171bacb306d7c7830a9cecb565b10bc618ad77e3c4c1c1893e8e82d6d0af7730ea73c7f000c27cb990c6c741b1377a` when app is closed and login - you will be redirected to user profile, on current nightly - to dashboard

#### Expected behavior
redirect to new chat `#tery`

#### Actual behavior
dashboard is shown

### Reproduction

- Tap on link https://get.status.im/chat/public/tery
- Login

### Additional Information

[comment]: # (Please do your best to fill this out.)
- Status version: nightly 21/08/2019

- Operating System: Android, iOS

#### Logs

Logs and video is [here](https://app.saucelabs.com/tests/03e0da2c9e114379ab03a007aa362e00?auth=998715fbab2df3d2fb1c08dbbde652dd#90)

### LINKED ISSUE #8830: "error No matching clause:offline" and white screen if try to open 1:1 chat [only on iOS]
### Description

[comment]: # (Feature or Bug? i.e Type: Bug)
*Type*: Bug

*Summary*: only on iOS: user can't open 1:1 chat because of `error No matching clause:offline`, after closing the popup app is not usable as it shows empty white screen, so relaunch is needed. After this 1:1 chat can't be opened again even if app is online. 

Assumption: To get the error user needs to move app on background and back while being offline but it's possible that issue happens in more cases

See exact steps below

#### Expected behavior

can open 1:1 chat and send/get messages

#### Actual behavior

"error No matching clause:offline"

### Reproduction

[comment]: # (Describe how we can replicate the bug step by step.)
video: https://drive.google.com/open?id=1J7SSLQ8b25dFPE1yD6I4T0Vq1Hl9u-aX
- Install Status
- Create new account
- select "maybe later" for both fingerprint and notifications
- add new contact by scanning its qr code (1:1 chat is opened, use Add to contacts link in the chat)
- send 1 text message in 1:1 chat
- close app
- launch app again
- tap "save password" and sign in
- Tap on "Allow" to allow notifications
- open 1:1 chat
- turn on Airplane mode, wait for "Offline" line to be shown in the app
- move app on background and back
- if no error shown then tap on Back to go to Chats and back to 1:1 chat 
- if no error shown try to move app to background couple of move times (also can scan another QR code with contact code to open new 1:1 chat)

### Additional Information

[comment]: # (Please do your best to fill this out.)
- Status version: nightly from Aug 23 https://ci.status.im/job/status-react/job/nightly/1223/
- Operating System:  iOS only: real iphone 6s iOS 12.4 and real iphone 7+ ios 12.3.1

#### Logs

[status_logs (5).zip](https://github.com/status-im/status-react/files/3535216/status_logs.5.zip)

### PR DESCRIPTION
**Pull request by yenda**
improves  #8759 Сan't send funds from added account (but you have to relogin first after creating the account, I asked Dmitry if he can fix it in the status-go PR here https://github.com/status-im/status-go/pull/1569)
fix #8760 multiple accounts created when tapping save account quickly (can't reproduce in this PR)
fix #8830 No matching clause:offline (accidently)
fix #8802 universal links after login (because why not)

This PR is a bit big because some cleanup was needed so that I could clearly see how to leverage some perf improvements (1 token balance call instead of all tokens * number of accounts) as well as implementation details (new accounts have to be saved via rpc call not just as a setting to make status-go aware of them) and integrate status-go fix for additional account transaction history before relogin. To sum it up

- clean up some mess with networks and current network being in multiple places in app-db
- clean up node status related code
- clean up transaction history fetching (remove all the event firing every block, only triggers when there is a transaction for the account)
- clean up balance fetching (replace multiple calls by a single one)
- use saveAccount rpc call so that new accounts get transaction history immediatly

### Reviewers

This is actually 12 separate commits that can be reviewed individually, each could be done in a separate PR but since they are kind of sequential it would have been a pita to make 12 PRs, 12 reviews and 12QAs

#### Areas that maybe impacted

- network related (tested changing network, adding/removing network, browser simpledapp.eth on testnet, wallet transactions)
- wallet transactions (tested transaction from initial wallet account to newly created wallet account right after creating new account to see if wallet history of the new account updates)

##### Functional

- wallet / transactions
- account recovery
- new account
- networks

##### Non-functional

- network consumption: might not consume less data but might be faster for your ens resolution in chat @asemiankevich because it uses a more efficient way to get token balances so less calls are made at startup

### Steps to test

Exploratory testing of the above mentionned areas. Already played a lot with it on both emulator and OnePlus5T real device

If you send eth transaction between you own accounts transfer might not be in the transaction list for one of the accounts and balance won't update until you relogin. This is a known issue that needs to be fixed in status-go

status: ready

### CODE REVIEW

File: src/status_im/multiaccounts/create/core.cljs [hunk 1]
```diff
@@ -142,15 +142,15 @@
   [multiaccount]
   [(let [{:keys [publicKey address]}
          (get-in multiaccount [:derived constants/path-default-wallet-keyword])]
-     {:public-key publicKey
```
Comment (flexsurfer) [inline src/status_im/multiaccounts/create/core.cljs:149]: maybe better to change this on go side? or implement normalization function for data from go side?

File: src/status_im/multiaccounts/create/core.cljs [hunk 2]
```diff
@@ -160,30 +160,32 @@
    {:keys [address publicKey keycard-instance-uid keycard-key-uid keycard-pairing keycard-paired-on mnemonic] :as multiaccount}
    password
    {:keys [seed-backed-up? login?] :or {login? true}}]
-  (let [[wallet-account {:keys [public-key]} :as subaccounts-data] (prepare-subaccounts-data multiaccount)
-        name (gfycat/generate-gfy public-key)
-        photo-path (identicon/identicon public-key)
+  (let [[wallet-account {:keys [publicKey]} :as subaccounts-data] (prepare-subaccounts-data multiaccount)
```
Comment (flexsurfer) [inline src/status_im/multiaccounts/create/core.cljs]: why subaccounts? not just account?

File: src/status_im/multiaccounts/login/core.cljs [hunk 3]
```diff
@@ -94,17 +92,58 @@
                          all-stored-browsers)]
     {:db (assoc db :browser/browsers browsers)}))
 
+(fx/defn initialize-web3-client-version
+  {:events [::initialize-web3-client-version]}
```
Comment (flexsurfer) [inline src/status_im/multiaccounts/login/core.cljs:96]: why do we need this?

File: src/status_im/wallet/core.cljs [hunk 4]
```diff
@@ -141,24 +146,49 @@
...
+                      acc)))
+                {}
+                results)]
+    (when (not-empty balances)
+      balances)))
+
 (re-frame/reg-fx
  :wallet/get-tokens-balances
  (fn [{:keys [addresses tokens assets]}]
    ;;TODO not great to have so many calls , should be optimized, there is wallet_getTokensBalances why wouldn't use it?
```
Comment (flexsurfer) [inline src/status_im/wallet/core.cljs]: so we can remove this todo now

**Review by flexsurfer — APPROVED** [review]
Eric is a real hacker! :)

File: src/status_im/multiaccounts/create/core.cljs (hunk 1 above)
Comment (yenda) [reply src/status_im/multiaccounts/create/core.cljs:149]: Yes homogenize the apis is one of my goal but here I just did what works

File: src/status_im/multiaccounts/create/core.cljs (hunk 2 above)
Comment (yenda) [reply src/status_im/multiaccounts/create/core.cljs]: Good point I'll change it tomorrow

File: src/status_im/multiaccounts/login/core.cljs (hunk 3 above)
Comment (yenda) [reply src/status_im/multiaccounts/login/core.cljs:96]: I don't know just for the about screen I guess? It is something that was there before.

**Comment by asemiankevich** [comment]
@yenda could you please double check if the fixes are in this branch?  i can repro all the issues listed in PR :(

for instance https://github.com/status-im/status-react/issues/8759 even with relogin
https://user-images.githubusercontent.com/28752642/63690318-fe213100-c814-11e9-93c4-e6dda346b6aa.png

or universal link, or multitap (on iOS)

https://user-images.githubusercontent.com/28752642/63694414-15651c00-c81f-11e9-8df8-71971e2c630e.png


https://user-images.githubusercontent.com/28752642/63694403-0d0ce100-c81f-11e9-9066-55131b1c7b36.png

**Comment by yenda** [comment]
@asemiankevich can you try on Android because I just retried all cases and it worked. Are you trying from fresh install? Is it the latest build?

**Comment by asemiankevich** [comment]
@yenda i am using  builds number 13 android 
https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-190825-000447-61d799-pr.apk
and number 12 iOS https://i.diawi.com/z6heGa

On android:

1. install status app and create new account
2. profile -> dev mode -> ropsten network
3. dapps -> simpledapp.eth -> faucet ETH
4. wallet -> add new account to multiaccount
5. wallet -> status account -> send some ETH to new account
6. switch to new account -> send some ETH back to Status account

I receive the `Transaction failed` message.

https://user-images.githubusercontent.com/28752642/63760083-59abf700-c8c7-11e9-8054-f2c0827fb29a.png


Also, on Android -> when adding new account to multiaccount:
1. open wallet -> add account 
2. enter password and multitap `Generate account` several times
3. multitap `Finish` button several times

as a result: several accounts added and error shown

https://user-images.githubusercontent.com/28752642/63760059-5153bc00-c8c7-11e9-9402-7d80f076eebb.png

**Comment by churik** [comment]
@asemiankevich 
Could reproduce issue with multitap on "Generate account" on my Android device.
Also is reproducible if tapping really quicky on IOS (could replicate only from 4 attempt).

But couldn't reproduce issur with `Transaction failed` message - tried on both IOS and Android couple of times

**Comment by yenda** [comment]
Ok I'll fix the multitap, for the transaction failed you need to logout and login after creating the new account and before sending the transaction, and also you need to send funds before otherwise balance may not update. These workarounds won't be necessary after status-go is fixed

**Comment by asemiankevich** [comment]
> Ok I'll fix the multitap, for the transaction failed you need to logout and login after creating the new account and before sending the transaction, and also you need to send funds before otherwise balance may not update. These workarounds won't be necessary after status-go is fixed

true, then i suggest not to close the issue https://github.com/status-im/status-react/issues/8759  until status go is fixed just in case

**Comment by churik** [comment]
@yenda also in this PR in addition to https://github.com/status-im/status-react/issues/8806 this error "null is not an object" is shown when creating new account.
I can add this to https://github.com/status-im/status-react/issues/8806 and fix it in terms of this issue.

**Comment by annadanchenko** [comment]
#8830 - can't reproduce anymore using iOS build 12 from this PR, so seems to be fixed (iphone 6s, iOS 12.4)

**Comment by asemiankevich** [comment]
@yenda seems i can't generate new account after incorrect password input attempt https://ethstatus.testrail.net/index.php?/tests/view/835215

Can you please take a look? I can do this in nightly fine, so steps:

1. install status
2. open wallet
3. tap Add account
4. enter wrong password and generate account -> validation error
5. re enter password to correct one and generate one more time

**Comment by flexsurfer** [comment]
https://github.com/status-im/status-go/pull/1569 has been merged

**Pull request merged**